Automattic/jetpack
# VideoPress: add debug() tool. Improve data-in-sync process

This PR uses the debug() tool to log actions in the dev console and also polishes and improves the video data-in-sync process.


Fixes https://github.com/Automattic/jetpack/issues/27570

# ⚙️ Data-in-sync process 

This is the name we refer to the process to keep in-sync the video data with the block through of its attributes.
Some video data fields should have the same value that block attributes, such as `title`, `description`, `privacy`, rating`, `tracks`, and so on. It means that we need to keep in-sync these data in both directions. From video data (provided by the server endpoint response body) to block attributes and vice versa.

## From data-video to block attributes

It happens when the block mounts. The app requests new video data, checks whether fields are out-of-sync, and sets the block attributes when needed.
Once the post saves, we consider block gets in sync with the video data.

### 🧪 Testing adding a new video block workflow

**Set your dev console to see the debug() lines**

* Open the dev console
* Set local storage debug item:
```
localStorage.setItem( 'debug', 'videopress:video:*' )
```

* probably, you may want to disable warns, errors, ... 
 
<img width="808" alt="Screen Shot 2022-11-30 at 07 35 39" src="https://user-images.githubusercontent.com/77539/204774031-60844880-6d28-44b3-ad00-4d2e179c08d3.png">

* Start with an empty post
* Add a new video block
* Upload the video
* After video uploads, confirm how the video data is propagated to the block attributes, taking a look at the dev console

<img width="987" alt="image" src="https://user-images.githubusercontent.com/77539/204777902-66bcabd9-bd70-4411-bd28-d9a770dd2ac2.png">

The logs above show how the app sets the block attributes based on the new video data. 

https://user-images.githubusercontent.com/77539/204777827-320309e1-fcf0-41fa-9e62-17cf3165bdc6.mov

### 🧪 Generate an out-of-sync state

* Edit the data of the same video used before, from the VideoPress dashboard, for instance, `title`, `description`, and `privacy`

`title` and `description` | `privacy`
------|------
<img width="1135" alt="Screen Shot 2022-11-30 at 07 59 07" src="https://user-images.githubusercontent.com/77539/204779315-9f4852c2-329a-46e8-9a92-af18dc7486b9.png"> | <img width="409" alt="Screen Shot 2022-11-30 at 08 00 03" src="https://user-images.githubusercontent.com/77539/204779303-8f84aa3b-1f7a-4ed7-8b64-8f6c288c4edd.png">

* Go back to the block editor where you added the video
* Confirm after a hard refresh, the app detects these fields are out-of-sync and updates the block attributes to keep them in sync

<img width="903" alt="Screen Shot 2022-11-30 at 08 05 21" src="https://user-images.githubusercontent.com/77539/204781043-ca5c2fcf-9cc6-4e87-bb19-cfacf0ef6438.png">
 again.

<img width="423" alt="Screen Shot 2022-11-30 at 08 06 45" src="https://user-images.githubusercontent.com/77539/204781035-2f754b1e-292a-4b20-9a69-84c08a7210b4.png">

* Save the post
* Hard refresh
* Confirm there is not any field out of sync.

## From block attributes to video data

The app watches and detects the data that needs to be updated to the video by hitting the proper endpoint. It happens when the user edits some block attributes, for instance.
The app collects all fields to update (title, description, ...) and performs a POST request right after the post saves, keeping the data in sync, but this time from the block to the video data.

### 🧪 Testing from block attributes to video-data

* Open the block sidebar
* Set `title`, `description`, ...
<img width="295" alt="Screen Shot 2022-11-30 at 08 21 52" src="https://user-images.githubusercontent.com/77539/204783588-b4115b5d-129e-44aa-a0ca-d841d740c003.png">

* Save the post
* Confirm the data detects these changes

<img width="899" alt="image" src="https://user-images.githubusercontent.com/77539/204783559-b224220d-341e-42dc-8c06-ea52de2dfa25.png">

* Confirm the app performs the POST request by taking a look at the dev console

<img width="903" alt="Screen Shot 2022-11-30 at 08 23 26" src="https://user-images.githubusercontent.com/77539/204785016-27cc3a81-76b3-4260-9989-fb957434f335.png">

* Go to the video details page of the VideoPress plugin
* Confirm the data changes you made have appropriately been propagated.

<img width="829" alt="Screen Shot 2022-11-30 at 08 30 46" src="https://user-images.githubusercontent.com/77539/204785236-4fecc4c9-840e-4df1-ac39-5abf48bfdf9e.png">

### 🧪 Testing auto-generated chapters

* Edit block description adding chapters
<img width="288" alt="Screen Shot 2022-11-30 at 08 33 17" src="https://user-images.githubusercontent.com/77539/204786253-99d634ad-c457-43b7-ace4-573415780d0f.png">

* Save the post
* Confirm the app detects and processes the chapters
<img width="899" alt="Screen Shot 2022-11-30 at 08 33 42" src="https://user-images.githubusercontent.com/77539/204786273-173309f9-e651-4175-b1b3-8cfad9539035.png">

* Confirm the app performs the request to upload the chapters
<img width="897" alt="Screen Shot 2022-11-30 at 08 36 46" src="https://user-images.githubusercontent.com/77539/204786481-1762af14-acd1-463f-b30b-44e36c12c76c.png">

* Confirm the player re-renders
* Confirm the player has chapters
<img width="523" alt="Screen Shot 2022-11-30 at 08 34 03" src="https://user-images.githubusercontent.com/77539/204786529-04af21a9-094f-4d7a-a04d-ebc5707076be.png">

* Confirm the chapter is the track's control dialog
<img width="474" alt="Screen Shot 2022-11-30 at 08 35 58" src="https://user-images.githubusercontent.com/77539/204786900-d27d7290-1d8b-4a4a-9b52-fb069305080b.png">


#### Changes proposed in this Pull Request:

* VideoPress: Add debug() tool to log actions in the dev console
* VideoPress: Improve video data-in-sync process

#### Jetpack product discussion



#### Does this pull request change what data or activity we track or use?



#### Testing instructions:

In the description
